haskell modules: Make 17.09 forwards compatible

4e6 opened the pull request:
###### Motivation for this change

Recent commit 81553124cf32e9d515da5f1c2f1b07e09886d8cf introduced new `buildHaskellPackages` argument and broke compatibility with _17.09_

Now if we call `makePackageSet` in _17.09_ with the new argument it fails with `called with unexpected argument ‘buildHaskellPackages’`.

This PR adds an ellipsis `(...)` to prevent `called with unexpected argument` errors in the future:

```
error: anonymous function at /nix/store/bfbxsw3jma333xv6x4r2kglwwa3f9zx8-nixpkgs-channels-c1d9aff56e0ae52ee4705440fe09291a51e91977-src/pkgs/development/haskell-modules/make-package-set.nix:4:1 called with unexpected argument ‘buildHaskellPackages’, at /nix/store/bfbxsw3jma333xv6x4r2kglwwa3f9zx8-nixpkgs-channels-c1d9aff56e0ae52ee4705440fe09291a51e91977-src/lib/customisation.nix:74:12
(use ‘--show-trace’ to show detailed location information)
```

cc @Ericson2314 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

4e6 committed:
Make haskell-modules backward compatible

Ericson2314 commented:
Uh shouldn't this be against 17.09?

Ericson2314 commented:
Also please add a comment on each ellipsis saying why. Otherwise looks good, thanks.

dhess commented:
@Ericson2314 I think the point here is to make sure that you can use `makePackageSet` with either `17.09` or a recent `nixpkgs`. Right now it's either-or because of the new parameter.

Ericson2314 commented:
@dhess yeah but to do that we need to allow passing the extra argument for both, and just have it be unused on, 17.09, so the extra unused parameter or ellipses (I actually prefer extra parameter, come to think of it, too) goes on 17.09.

dhess commented:
@Ericson2314 OK, thanks for the clarification. (I am not the author but am eagerly awaiting the fix ;)

peti commented on pkgs/development/haskell-modules/default.nix:
```diff
@@ -6,6 +6,7 @@
 , initialPackages ? import ./initial-packages.nix
 , configurationCommon ? import ./configuration-common.nix
 , configurationNix ? import ./configuration-nix.nix
+, ...
```
Please do not add ellipsis here. I don't want these functions to accept any random argument without an error.

4e6 replied:
ok, I'll use `makePackgeSet` instead

peti replied:
I am sorry, but that is not what I meant. Please don't accept `...` arguments *at all*. I don't want any of these functions to take random arguments without an error.

4e6 committed:
Add comments

4e6 commented:
I added the comments. and addressed @peti remark

Ellipsis is added to create (some kind of) backward compatibility. With ellipsis, when argument list is extended in _master_, we can still support both _master_ and _release_ just by passing extra arguments.

Ericson2314 commented:
Yeah could you do `buildHaskellPackages ? null` instead?

4e6 commented:
> Yeah could you do buildHaskellPackages ? null instead?

But it is a required argument, it should never be null. What's wrong with the ellipsis?

Ericson2314 commented:
@4e6 you missed me saying this PR should point at `release-17.09` instead.

Ericson2314 commented:
I went ahead and changed the branch for you.

peti reviewed (changes requested):
Please accept specific arguments rather than using `...`.

You also need to rebase your PR relative to the current `release-17.09` branch.

4e6 committed:
Replace ellipsis with null argument

4e6 commented:
I didn't get that we are talking about merging into `release-17.09`. Then `buildHaskellPackages ? null` makes sense. Sorry for the noise.

Ericson2314 commented:
Looks great now. Thanks!

Ericson2314 commented:
Oops meant to squash, but it didn't do it. Oh well.
